Exception encountered when running odo link on minikube

/kind bug

<!--

Welcome! - We kindly ask you to:

  1. Fill out the issue template below 
  2. Use the Google group if you have a question rather than a bug or feature request.

The group is at: https://groups.google.com/forum/#!forum/odo-users

Thanks for understanding, and for contributing to the project!

-->

## What versions of software are you using?

**Operating System:**
Distributor ID:	Ubuntu
Description:	Ubuntu 18.04.5 LTS
Release:	18.04
Codename:	bionic

Kubernetes v1.20.2 on Docker 20.10.2

**Output of `odo version`:**
odo v2.0.4 (5544b25ec)

## How did you run odo exactly?
After listing an available service:
```
root@unwisest1:~/odoprojects/application-stack-samples/jpa# odo service list
NAME                   AGE
Database/database3     6m35s
```

I then attempted  to link to that service:
`root@unwisest1:~/odoprojects/application-stack-samples/jpa# odo link Database/database3`

## Actual behavior
Exception is raised when issuing the odo link.

## Expected behavior
odo link is successful

## Any logs, error output, etc?
Exception raised:
```
root@unwisest1:~/odoprojects/application-stack-samples/jpa# odo link Database/database3
panic: runtime error: index out of range [2] with length 1
goroutine 1 [running]:
github.com/openshift/odo/pkg/odo/cli/component.(*commonLinkOptions).complete(0xc000305680, 0xc0006a0510, 0x4, 0xc00019b180, 0xc00076e290, 0x1, 0x1, 0x1682b9a, 0xc0006a0510)
	/builddir/build/BUILD/gocode/src/github.com/openshift/odo/pkg/odo/cli/component/common_link.go:138 +0xfe6
github.com/openshift/odo/pkg/odo/cli/component.(*LinkOptions).Complete(0xc0006b35c0, 0xc0006a0510, 0x4, 0xc00019b180, 0xc00076e290, 0x1, 0x1, 0xc000140800, 0x1e66f93)
	/builddir/build/BUILD/gocode/src/github.com/openshift/odo/pkg/odo/cli/component/link.go:100 +0x121
github.com/openshift/odo/pkg/odo/genericclioptions.GenericRun(0x2208960, 0xc0006b35c0, 0xc00019b180, 0xc00076e290, 0x1, 0x1)
	/builddir/build/BUILD/gocode/src/github.com/openshift/odo/pkg/odo/genericclioptions/runnable.go:29 +0xb8
github.com/openshift/odo/pkg/odo/cli/component.NewCmdLink.func1(0xc00019b180, 0xc00076e290, 0x1, 0x1)
	/builddir/build/BUILD/gocode/src/github.com/openshift/odo/pkg/odo/cli/component/link.go:158 +0x5e
github.com/spf13/cobra.(*Command).execute(0xc00019b180, 0xc00076e270, 0x1, 0x1, 0xc00019b180, 0xc00076e270)
	/builddir/build/BUILD/gocode/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:830 +0x29d
github.com/spf13/cobra.(*Command).ExecuteC(0xc0003a6280, 0x1fac418, 0xc00010efc0, 0x2)
	/builddir/build/BUILD/gocode/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:914 +0x2fb
github.com/spf13/cobra.(*Command).Execute(...)
	/builddir/build/BUILD/gocode/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:864
main.main()
	/builddir/build/BUILD/gocode/src/github.com/openshift/odo/cmd/odo/odo.go:67 +0x32a
root@unwisest1:~/odoprojects/application-stack-samples/jpa#
```

Thanks Dharmit - 

Based on your success, I got myself a clean linux vm and tried this again from the beginning, but I continue to hit the exception on the new machine as well. I don't know what might be broken in my process, but still cannot get link to work against minikube on linux:

```Service "Database/sampledatabase" has been successfully deleted
[root@pappuses1 jpa]# odo service create postgresql-operator.v0.1.1/Database
Deploying service of type: Database
 ✓  Deploying service [8ms]
 ✓  Service 'database' was created

Progress of the provisioning will not be reported and might take a long time
You can see the current status by executing 'odo service list'
Optionally, link postgresql-operator.v0.1.1 to your component by running: 'odo link <component-name>'
[root@pappuses1 jpa]# odo link Database/database
panic: runtime error: index out of range [2] with length 1

goroutine 1 [running]:
github.com/openshift/odo/pkg/odo/cli/component.(*commonLinkOptions).complete(0xc0005c6960, 0xc00016a480, 0x4, 0xc0005e0500, 0xc0008920d0, 0x1, 0x1, 0x180348a, 0xc00016a480)
	/builddir/build/BUILD/gocode/src/github.com/openshift/odo/pkg/odo/cli/component/common_link.go:138 +0xfe6
github.com/openshift/odo/pkg/odo/cli/component.(*LinkOptions).Complete(0xc000410fe0, 0xc00016a480, 0x4, 0xc0005e0500, 0xc0008920d0, 0x1, 0x1, 0xc000622100, 0x1e688b3)
	/builddir/build/BUILD/gocode/src/github.com/openshift/odo/pkg/odo/cli/component/link.go:100 +0x121
github.com/openshift/odo/pkg/odo/genericclioptions.GenericRun(0x2207e00, 0xc000410fe0, 0xc0005e0500, 0xc0008920d0, 0x1, 0x1)
	/builddir/build/BUILD/gocode/src/github.com/openshift/odo/pkg/odo/genericclioptions/runnable.go:29 +0xb8
github.com/openshift/odo/pkg/odo/cli/component.NewCmdLink.func1(0xc0005e0500, 0xc0008920d0, 0x1, 0x1)
	/builddir/build/BUILD/gocode/src/github.com/openshift/odo/pkg/odo/cli/component/link.go:158 +0x5e
github.com/spf13/cobra.(*Command).execute(0xc0005e0500, 0xc0008920b0, 0x1, 0x1, 0xc0005e0500, 0xc0008920b0)
	/builddir/build/BUILD/gocode/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:830 +0x29d
github.com/spf13/cobra.(*Command).ExecuteC(0xc000164c80, 0x1faddf8, 0xc0008a6000, 0x2)
	/builddir/build/BUILD/gocode/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:914 +0x2fb
github.com/spf13/cobra.(*Command).Execute(...)
	/builddir/build/BUILD/gocode/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:864
main.main()
	/builddir/build/BUILD/gocode/src/github.com/openshift/odo/cmd/odo/odo.go:67 +0x32a
[root@pappuses1 jpa]# ```

Reproducible on 4.8. Makes case for having https://github.com/openshift/odo/issues/4503 stronger!
```sh
$ odo link EtcdCluster/example
panic: runtime error: index out of range [2] with length 1

goroutine 1 [running]:
github.com/openshift/odo/pkg/odo/cli/component.(*commonLinkOptions).complete(0xc000788f00, 0xc00067c480, 0x4, 0xc000790500, 0xc0007d30f0, 0x1, 0x1, 0x162ac5a, 0xc00067c480)
	/home/dshah/go/src/github.com/openshift/odo/pkg/odo/cli/component/common_link.go:142 +0x1029
github.com/openshift/odo/pkg/odo/cli/component.(*LinkOptions).Complete(0xc000644f20, 0xc00067c480, 0x4, 0xc000790500, 0xc0007d30f0, 0x1, 0x1, 0xc000798300, 0x1f793f6)
	/home/dshah/go/src/github.com/openshift/odo/pkg/odo/cli/component/link.go:100 +0x121
github.com/openshift/odo/pkg/odo/genericclioptions.GenericRun(0x2332a60, 0xc000644f20, 0xc000790500, 0xc0007d30f0, 0x1, 0x1)
	/home/dshah/go/src/github.com/openshift/odo/pkg/odo/genericclioptions/runnable.go:29 +0xb8
github.com/openshift/odo/pkg/odo/cli/component.NewCmdLink.func1(0xc000790500, 0xc0007d30f0, 0x1, 0x1)
	/home/dshah/go/src/github.com/openshift/odo/pkg/odo/cli/component/link.go:158 +0x5e
github.com/spf13/cobra.(*Command).execute(0xc000790500, 0xc0007d30d0, 0x1, 0x1, 0xc000790500, 0xc0007d30d0)
	/home/dshah/go/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:830 +0x2aa
github.com/spf13/cobra.(*Command).ExecuteC(0xc000670c80, 0x20c7830, 0xc000056360, 0x2)
	/home/dshah/go/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:914 +0x2fb
github.com/spf13/cobra.(*Command).Execute(...)
	/home/dshah/go/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:864
main.main()
	/home/dshah/go/src/github.com/openshift/odo/cmd/odo/odo.go:67 +0x330
```

--- a/pkg/kclient/deployments.go
+++ b/pkg/kclient/deployments.go
@@ -383,3 +383,31 @@ func (c *Client) GetDeploymentConfigsFromSelector(selector string) ([]appsv1.Dep
 	}
 	return dcList.Items, nil
 }
+
+// GetDeploymentAPIVersion returns a map with Group, Version, Resource information of Deployment objects
+// depending on the GVR supported by the cluster
+func (c *Client) GetDeploymentAPIVersion() (map[string]string, error) {
+	extV1Beta1, err := c.IsDeploymentExtensionsV1Beta1()
+	if err != nil {
+		return nil, err
+	}
+
+	if extV1Beta1 {
+		// this indicates we're running on OCP 3.11 cluster
+		return map[string]string{
+			"group":    "extensions",
+			"version":  "v1beta1",
+			"resource": "deployments",
+		}, nil
+	}
+
+	return map[string]string{
+		"group":    "apps",
+		"version":  "v1",
+		"resource": "deployments",
+	}, nil
+}
+
+func (c *Client) IsDeploymentExtensionsV1Beta1() (bool, error) {
+	return c.IsResourceSupported("extensions", "v1beta1", "deployments")
+}

--- a/pkg/kclient/deployments.go
+++ b/pkg/kclient/deployments.go
@@ -386,25 +386,25 @@ func (c *Client) GetDeploymentConfigsFromSelector(selector string) ([]appsv1.Dep
 
 // GetDeploymentAPIVersion returns a map with Group, Version, Resource information of Deployment objects
 // depending on the GVR supported by the cluster
-func (c *Client) GetDeploymentAPIVersion() (map[string]string, error) {
+func (c *Client) GetDeploymentAPIVersion() (metav1.GroupVersionResource, error) {
 	extV1Beta1, err := c.IsDeploymentExtensionsV1Beta1()
 	if err != nil {
-		return nil, err
+		return metav1.GroupVersionResource{}, err
 	}
 
 	if extV1Beta1 {
 		// this indicates we're running on OCP 3.11 cluster
-		return map[string]string{
-			"group":    "extensions",
-			"version":  "v1beta1",
-			"resource": "deployments",
+		return metav1.GroupVersionResource{
+			Group:    "extensions",
+			Version:  "v1beta1",
+			Resource: "deployments",
 		}, nil
 	}
 
-	return map[string]string{
-		"group":    "apps",
-		"version":  "v1",
-		"resource": "deployments",
+	return metav1.GroupVersionResource{
+		Group:    "apps",
+		Version:  "v1",
+		Resource: "deployments",
 	}, nil
 }
 

--- a/pkg/odo/cli/component/common_link.go
+++ b/pkg/odo/cli/component/common_link.go
@@ -120,20 +120,18 @@ func (o *commonLinkOptions) complete(name string, cmd *cobra.Command, args []str
 		// make this deployment the owner of the link we're creating so that link gets deleted upon doing "odo delete"
 		ownerReference := generator.GetOwnerReference(deployment)
 		o.serviceBinding.SetOwnerReferences(append(o.serviceBinding.GetOwnerReferences(), ownerReference))
+
+		deploymentGVR, err := o.KClient.GetDeploymentAPIVersion()
 		if err != nil {
 			return err
 		}
 
-		// This is a really hacky way to get group, version and resource info but I couldn't find better one.
-		// A sample "deploymentSelfLinkSplit" looks like: [ apis apps v1 namespaces myproject deployments nodejs ]
-		deploymentSelfLinkSplit := strings.Split(deployment.SelfLink, "/")
-
 		// Populate the application selector field in service binding request
 		o.serviceBinding.Spec.Application = &servicebinding.Application{
 			GroupVersionResource: metav1.GroupVersionResource{
-				Group:    deploymentSelfLinkSplit[2], // "apps" in above example output
-				Version:  deploymentSelfLinkSplit[3], // "v1" in above example output
-				Resource: deploymentSelfLinkSplit[6], // "deployments" in above example output
+				Group:    deploymentGVR["group"],
+				Version:  deploymentGVR["version"],
+				Resource: deploymentGVR["resource"],
 			},
 		}
 

--- a/pkg/odo/cli/component/common_link.go
+++ b/pkg/odo/cli/component/common_link.go
@@ -129,9 +129,9 @@ func (o *commonLinkOptions) complete(name string, cmd *cobra.Command, args []str
 		// Populate the application selector field in service binding request
 		o.serviceBinding.Spec.Application = &servicebinding.Application{
 			GroupVersionResource: metav1.GroupVersionResource{
-				Group:    deploymentGVR["group"],
-				Version:  deploymentGVR["version"],
-				Resource: deploymentGVR["resource"],
+				Group:    deploymentGVR.Group,
+				Version:  deploymentGVR.Version,
+				Resource: deploymentGVR.Resource,
 			},
 		}
 

--- a/pkg/odo/cli/component/common_link.go
+++ b/pkg/odo/cli/component/common_link.go
@@ -128,11 +128,7 @@ func (o *commonLinkOptions) complete(name string, cmd *cobra.Command, args []str
 
 		// Populate the application selector field in service binding request
 		o.serviceBinding.Spec.Application = &servicebinding.Application{
-			GroupVersionResource: metav1.GroupVersionResource{
-				Group:    deploymentGVR.Group,
-				Version:  deploymentGVR.Version,
-				Resource: deploymentGVR.Resource,
-			},
+			GroupVersionResource: deploymentGVR,
 		}
 
 		o.serviceBinding.Spec.Application.Name = componentName

